### DISCUSSION TITLE
Get signable content from and add the signed content to a PDF

### DISCUSSION THREAD

**Discussion by carlosalbertos** [empira/PDFsharp#178] [Q&A] [answered]
Hello PDFsharp team!
I need to obtain a digitally signed PDF file with a particularity. The signature is not processed locally, since I do not have the certificate's pfx file. The signature is generated by a cloud-based digital certificate API.

For this API I need to pass the digest of the PDF file content, SHA256 of the Base64-encoded content. Here comes the first question. How do I obtain this content through PDFsharp?

The API will return the signed content in Base64 format (+timestamp, + encodedX509 and fingerprint256 from certificate) . Second question: How do I add the signature fields to the original PDF using the PDFsharp library methods?

I have consulted all the information, examples, issues, including the source code, but I have not found an answer for my case.

Can PDFsharp 1.5 or 6.2.0 Preview 1 provide these functionalities? Do you have an example code to share?
Thank you in advance.

**Comment by ThomasHoevel** [comment, marked as answer]
Look at the signature samples:
https://github.com/empira/PDFsharp.Samples/tree/master/src/samples/src/PDFsharp/src/Pdf.Signatures

Implement your own instance of the IDigitalSigner interface and make it work with your cloud-based API.

**Reply by carlosalbertos** [reply]
Thanks for the reply Thomas. I'll try.

**Comment by TM-Atharva** [comment]
This is Awesome!!

How to get multiple sign on same and multiple pages?

Should I run this code in loop?

```var cert = new X509Certificate2(@"C:\Test Digital Certificate Password is 123456.pfx", "123456");
var options = new DigitalSignatureOptions
{
   

    ContactInfo = "SignMe",
    Location = "World",
    Reason = "License Agreement",
    Rectangle = new XRect(120, 40, 100, 60),
    AppearanceHandler = new DefaultSigner.SignatureAppearanceHandler()

    // Signature appearances can also consist of an image (Rectangle should be adapted to image's aspect ratio)
    //Image = XImage.FromFile(@"C:\Data\stamp.png")
};

// Specify the URI of a timestamp server if you want a signature with timestamp.
var pdfSignatureHandler = DigitalSignatureHandler.ForDocument(finalDoc,
    new PdfSharpDefaultSigner(cert, PdfMessageDigestType.SHA256, new Uri("http://timestamp.apple.com/ts01")),
    options);`
    
    Regards,

**Comment by ThomasHoevel** [comment]
Multiple signatures are not yet supported. This requires incremental updates - you can try the code from the PR, but this is not yet part of PDFsharp.

**Reply by TM-Atharva** [reply]
Hey @ThomasHoevel - Thanks for the quick reply!

out of curiosity  - Any reason why that [PR ](https://github.com/empira/PDFsharp/pull/112) is not merged or any other plan ?

**Reply by TM-Atharva** [reply]
Hey @ThomasHoevel  / @PDFsharp-Team 

I have successfully signed document on multi page / multi places.

Do you know how to set permissions on finally signed document as once we open signed document for setting permissions then signs are getting invalid certificate errors?

**Comment by TH-Soft** [comment]
The signature verifies that the PDF was not tampered with. Any change to the PDF after signing it invalidates the signature.
Make all other changes before signing it.

**Reply by TM-Atharva** [reply]
So you mean, we should set permission like below first and then send for signer to sign the document?

```
finalDoc.SecuritySettings.OwnerPassword = "ownpwd";
finalDoc.SecuritySettings.PermitAnnotations = false;
finalDoc.SecuritySettings.PermitAssembleDocument = false;
finalDoc.SecuritySettings.PermitExtractContent = false;
finalDoc.SecuritySettings.PermitFormsFill = false;
finalDoc.SecuritySettings.PermitFullQualityPrint = false;
finalDoc.SecuritySettings.PermitModifyDocument = false;
finalDoc.SecuritySettings.PermitPrint = true;
```

I have tried this all but when I sign document, it ask for password so I did as below 

`document = PdfReader.Open(filename, ownerPassword, PdfDocumentOpenMode.Modify);`

but still we are getting error that its password protected PDF and needs owner password.

UseCase: Send document for sing to multiple user and send final signed copy to everyone with protected permission.

**Comment by TH-Soft** [comment]
Are you sure you used the correct owner password?

It's better to combine as many steps as possible for each "Save" call. Ideally you only have to call "Save" once.

Option 1: Open the document, add the security settings, add the signature, save it.

Option 2: If owner password was already set, open PDF with owner password, add the signature, save it.

**Reply by TM-Atharva** [reply]
Both the options we tried, but if PDF is just for one page it works but for multiple pages to sign and set security settings are not working together. 

@ThomasHoevel We are making one product on the base of PDFSharp so we are heavily dependent on this Library.

Pls suggest!

**Reply by TM-Atharva** [reply]
Hey @ThomasHoevel  / @TH-Soft  !

!! I have successfully managed to set permissions on the final signed copy of the PDF !!

I have used incremental sign approach as per that PR and able to accomplish given task.  I am converting PDF to images for multiple sign coordinates. 

My only challenge now it how can I keep text searchable in PDF even after sign and remove PDF to Image approach.

Thank you very much for your support.

**Reply by Havunen** [reply]
@TM-Atharva  

Did you fork the library to implement that functionality? Do you have it open sourced somewhere?

**Reply by TM-Atharva** [reply]
@Havunen  I have just used that [PR ](https://github.com/empira/PDFsharp/pull/112) repo for signing. Security and Permissions on signed document is still pending.
